Do not restart FTL while `pihole -g` is still ongoing

DL6ER opened the pull request:
# What does this implement/fix?

Do not try to restart FTL when a streamed `pihole -g` run is still ongoing. This to prevent being stuck in a situation where FTL has shut down 99% and cannot serve DNS until `pihole -g` finished running.

---

**Related issue or feature (if applicable):** Fixes #2395 

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** N/A

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `developmental` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

yubiuser commented:
Why are we offering the `gravity` endpoint at all? Because we want to be able to run gravity from the web interface?

___

Im not sure if this is the right way to fix https://github.com/pi-hole/FTL/issues/2395 - maybe the external program should simply call `pihole -g` instead of using the API endpoint?

DL6ER commented:
> Why are we offering the gravity endpoint at all? Because we want to be able to run gravity from the web interface?

Yes

> Im not sure if this is the right way to fix https://github.com/pi-hole/FTL/issues/2395 - maybe the external program should simply call pihole -g instead of using the API endpoint?

Yes, but my point here is that - as long as we are offering this - this can very well be considered a bug as FTL can be brought into a (temporary, but maybe long) defunct state until gravity finishes it can finally restart

yubiuser reviewed (commented):
It causes some logspam

```
2025-04-06 21:31:59.322 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:00.323 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:01.322 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:02.323 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:03.324 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:04.325 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:05.325 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:06.326 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:07.327 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:08.328 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:09.328 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:10.329 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:11.329 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:12.330 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:13.332 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:13.641 CEST [135290/T135558] INFO: Restarting FTL: API action request
2025-04-06 21:32:13.642 CEST [135290M] INFO: Asked to terminate by "/usr/bin/pihole-FTL -f" (PID 135290, user pihole UID 999)
2025-04-06 21:32:13.642 CEST [135290M] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:14.331 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:15.332 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:16.332 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
2025-04-06 21:32:17.333 CEST [135290/T135309] INFO: Not restarting as gravity is still running...
```

__

This is working as expected, but something very 'gravity' specific. Shouldn't this be extended to all endpoints - only restart when the previous requests have been finished?

DL6ER commented:
Sorry, seeing your reply only now. The gravity case is very special here because it is (really!) long-lived and can prevent FTL from shutting down. Is there any other action that is capable of doing something similar? I don't think so.

We can surely reduce the logging to do it ... once and then again every 30 seconds if still waiting? ...

yubiuser commented:
> We can surely reduce the logging to do it ... once and then again every 30 seconds if still waiting? ...

Yes please. And more importantly: Only start logging when gravity is running and a restart request has been received. As you can see in the log above the logging started when the gravity endpoint has been requested, and only 14 seconds later the restart requests was received.

yubiuser commented:
I think this needs some attention from @DL6ER. I added some review points

> Yes please. And more importantly: Only start logging when gravity is running and a restart request has been received. As you can see in the log above the logging started when the gravity endpoint has been requested, and only 14 seconds later the restart requests was received.

DL6ER committed:
Print gravity running warning only when wanting to terminate and at most every 30 seconds

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Do not try to restart FTL when a streamed pihole -g run is still ongoing. This to prevent being stuck in a situation where FTL has shut down 99% and cannot serve DNS until pihole -g finished running

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Merge branch 'development' into tweak/prevent_restart_during_gravity_run

DL6ER committed:
Merge branch 'development' into tweak/prevent_restart_during_gravity_run

Signed-off-by: Dominik <dl6er@dl6er.de>

yubiuser reviewed (changes requested):
I'm not sure how I tested it last time, but now I tried it with two approaches and it killed `gravity` right away without printing any warning or waiting for `gravity`.

1. Triggered gravity from the web interface, then re-started resolver from Settings/System.
2. Triggered gravity from the web interface, then changed a setting that required a FTL restart

DL6ER commented:
@yubiuser Did you meant to submit this as request for changes? If so, I'm not seeing an actual request.

yubiuser commented:
Yes it is. The request is to make it work. Currently, a running gravity is killed without waiting when a FTL restart is triggered

DL6ER committed:
Fix logic

Signed-off-by: Dominik <dl6er@dl6er.de>

yubiuser reviewed (dismissed):
```
2025-10-05 11:50:32.551 CEST [172358/T172828] INFO: Restarting FTL: API action request
2025-10-05 11:50:32.552 CEST [172358M] INFO: Asked to terminate by "/usr/bin/pihole-FTL -f" (PID 172358, user pihole UID 999)
2025-10-05 11:50:32.552 CEST [172358M] INFO: Not terminating as gravity is still running...
2025-10-05 11:51:03.959 CEST [172358/T172818] INFO: Not terminating as gravity is still running...
2025-10-05 11:51:34.975 CEST [172358/T172818] INFO: Not terminating as gravity is still running...
2025-10-05 11:51:42.033 CEST [172358/T172820] INFO: Terminating timer thread
2025-10-05 11:51:42.044 CEST [172358/T172817] INFO: Terminating database thread
2025-10-05 11:51:42.329 CEST [172358M] INFO: Finished final database update
2025-10-05 11:51:42.330 CEST [172358M] INFO: Waiting for threads to join
2025-10-05 11:51:42.330 CEST [172358M] INFO: Thread housekeeper (1) is idle, terminating it.
2025-10-05 11:51:42.330 CEST [172358M] INFO: Thread dns-client (2) is idle, terminating it.

```

DL6ER committed:
Merge branch 'development' into tweak/prevent_restart_during_gravity_run

DL6ER committed:
Merge branch 'development' into tweak/prevent_restart_during_gravity_run

Signed-off-by: Dominik <dl6er@dl6er.de>

yubiuser reviewed (commented):
I tested it again with the following setting:
Triggering gravity from the web interface plus....

1) send restart via API (`/action/restartdns`) -> restart was delayed
2) changed a setting via web interface that required a restart -> restart was delayed
3) restated FTL via `sudo systemctl restart pihole-FTL.service` -> restarted immediately

I'm not sure if 3) is expected. It logged that it would not restart, but did anyway.

```
2026-02-19 15:34:43.403 UTC [2005424M] INFO: Asked to terminate by "/sbin/init" (PID 1, user root UID 0)
2026-02-19 15:34:43.404 UTC [2005424M] INFO: Not terminating as gravity is still running...
2026-02-19 15:34:43.411 UTC [2005424/T2008378] ERROR: gravity failed with signal 15 
2026-02-19 15:34:43.412 UTC [2005424/T2008378] WARNING: API: Gravity failed (key: server_error)

```

yubiuser commented on src/gc.c:
```diff
@@ -710,6 +710,14 @@ void *GC_thread(void *val)
+
+		// Check if we need to terminate/restart FTL but this has been
+		// postponed because of an ongoing gravity run
+		check_if_want_terminate();
```
How do you reach this part of the code? As `want_terminate` is not set set, `check_if_want_terminate()` will always return immediately.

rdwebdesign replied:
`want_terminate` is set in `daemon.c` line 59.

DL6ER committed:
Merge branch 'development' into tweak/prevent_restart_during_gravity_run

DL6ER committed:
Detach the forked child into its own session so SIGTERM (15) and other signals sent to the parent's process group won't reach it causing a possible premature disruption

Signed-off-by: Dominik <dl6er@dl6er.de>

DL6ER commented:
(3) should be fixed as well now

DL6ER committed:
fix: use volatile sig_atomic_t for gravity_running/want_terminate and trigger deferred restart immediately

gravity_running and want_terminate are both read and written from
SIGTERM_handler(), a signal handler context. Plain bool is not safe
here: the C standard (§7.14.1.1) only guarantees defined behavior for
volatile sig_atomic_t (or lock-free atomics) when accessed from signal
handlers. A compiler is free to cache a plain bool in a register across
a signal delivery boundary.

Also call check_if_want_terminate() directly after clearing
gravity_running in api_action_gravity(), so a deferred restart is acted
on the moment gravity finishes rather than waiting up to ~1 s for the
next GC thread iteration.

Signed-off-by: Dominik <dl6er@dl6er.de>

DL6ER committed:
Merge branch 'development' into tweak/prevent_restart_during_gravity_run

DL6ER commented:
I forgot about this PR meanwhile and was already working on a new/similar fix. I pushed a minor change to this PR (and updated it to latest `development`) but it would also have been fine to merge the previous state (`bool` is atomic on platforms being relevant for us).

yubiuser commented:
3) restated FTL via `sudo systemctl restart pihole-FTL.service` -> restarted immediately

Still kills FTL right away.

```
2026-03-22 08:29:58.235 UTC [1306679M] INFO: Asked to terminate by "/usr/lib/systemd/systemd --system --deserialize=46" (PID 1, user root UID 0)
2026-03-22 08:29:58.235 UTC [1306679M] INFO: Not terminating as gravity is still running...
2026-03-22 08:29:58.244 UTC [1306679/T1306846] ERROR: gravity failed with signal 15 
2026-03-22 08:29:58.244 UTC [1306679/T1306846] WARNING: API: Gravity failed (key: server_error)
2026-03-22 08:29:58.296 UTC [1306679/T1306681] INFO: Terminating database thread
2026-03-22 08:29:58.304 UTC [1306679/T1306683] INFO: Terminating resolver thread
2026-03-22 08:29:58.333 UTC [1306679/T1306684] INFO: Terminating timer thread
2026-03-22 08:29:58.514 UTC [1306679M] INFO: Finished final database update

```

DL6ER committed:
fix: ignore SIGTERM in gravity child to survive systemd cgroup-level kill

setsid() was already added to detach the child from the parent's
process group, but systemd's default KillMode=control-group sends
SIGTERM to ALL processes in the cgroup, not just the process group.
After execv(), custom signal handlers are reset to SIG_DFL, so the
gravity child was still killed immediately.

Set SIG_IGN for SIGTERM before execv() since SIG_IGN (unlike custom
handlers) is preserved across exec. This lets gravity run to
completion even when FTL receives SIGTERM from systemd.

Signed-off-by: Dominik <dl6er@dl6er.de>

DL6ER commented:
This was quite some rabbit hole...

The root cause is that `systemd`'s default `KillMode=control-group` sends `SIGTERM` to all processes in the `cgroup`, not just the main PID. FTL's `SIGTERM` handler correctly defers shutdown ("Not terminating as gravity is still running..."), but the gravity child process - which runs `pihole -g` via `execv()` - has default `SIGTERM` disposition because `execv()` resets all custom signal handlers to `SIG_DFL`. So the child gets killed by the `cgroup`-wide `SIGTERM` just 9ms after the parent defers.                                                          

Fixed now (hopefully!) by adding `setsid()` and `signal(SIGTERM, SIG_IGN)` in the child process before `execv()`. `SIG_IGN` (unlike custom handlers) is preserved across `exec`, so the gravity script and its children will ignore `SIGTERM` from above. This lets gravity complete naturally, after which FTL proceeds with the deferred shutdown.

Could you please re-test? I am not sure why I am not able to reproduce your finding on my test setup.

yubiuser reviewed (approved):
Works fine now. Thanks for digging through.

```
2026-03-29 10:28:02.227 UTC [399861M] INFO: Asked to terminate by "/sbin/init" (PID 1, user root UID 0)
2026-03-29 10:28:02.228 UTC [399861M] INFO: Not terminating as gravity is still running...
2026-03-29 10:28:02.906 UTC [399861/T403021] INFO: Received 8/8 valid NTP replies from de.pool.ntp.org
2026-03-29 10:28:02.906 UTC [399861/T403021] INFO: NTP time offset: -3.194639e+00 ms (excluded 1 outliers)
2026-03-29 10:28:02.906 UTC [399861/T403021] INFO: NTP round-trip delay: 2.931683e+01 ms (excluded 1 outliers)
2026-03-29 10:28:03.000 UTC [399861/T403021] INFO: Current RTC time is 2026-03-29 10:28:03
2026-03-29 10:28:33.671 UTC [399861/T403023] INFO: Not terminating as gravity is still running...
2026-03-29 10:28:44.342 UTC [399861/T403025] INFO: Terminating timer thread
2026-03-29 10:28:44.375 UTC [399861/T403022] INFO: Terminating database thread
2026-03-29 10:28:44.591 UTC [399861M] INFO: Finished final database update
2026-03-29 10:28:44.591 UTC [399861M] INFO: Waiting for threads to join
2026-03-29 10:28:44.591 UTC [399861M] INFO: Thread housekeeper (1) is idle, terminating it.
2026-03-29 10:28:44.591 UTC [399861M] INFO: Thread dns-client (2) is idle, terminating it.
2026-03-29 10:28:44.592 UTC [399861M] INFO: Thread ntp-client (4) is idle, terminating it.
2026-03-29 10:28:44.592 UTC [399861M] INFO: Thread webserver (7) is idle, terminating it.
2026-03-29 10:28:44.592 UTC [399861M] INFO: All threads joined
2026-03-29 10:28:44.593 UTC [399861M] INFO: PID file emptied
2026-03-29 10:28:44.595 UTC [399861M] INFO: Stored 0 API sessions in the database
2026-03-29 10:28:44.686 UTC [399861M] INFO: ########## FTL terminated after 56s  (code 0)! ##########

```

yubiuser commented:
Oh it was set to auto-merge... we should have squashed the the 'merge development' commits.
